images/dnsmasq/Makefile

```diff
@@ -89,74 +82,56 @@ all-containers: $(addprefix containers-, $(ALL_ARCH))
-		-v `pwd`:/src                 \
-		$(COMPILE_IMAGE)              \
-		/bin/sh -c                    \
-		"apk update                                 \
-			&& apk add alpine-sdk xz libcap-dev \
-			&& tar -xJf /build/dnsmasq.tar.xz   \
-			&& cd $(DNSMASQ_VERSION)            \
-			&& make -j                          \
-			&& cp src/dnsmasq /build" $(VERBOSE_OUTPUT)
-else
-	@cd $(OUTPUT_DIR)                                            \
-		&& sed -i "s|__ARCH__|$(QEMUARCH)|g" Dockerfile      \
-		&& sed -i "s/__CROSS_BUILD_COPY__/COPY/g" Dockerfile
-	@cd $(OUTPUT_DIR) \
-		&& curl -sSL $(MULTIARCH_RELEASE) | tar -xJ
-	@docker run --sig-proxy=true --rm            \
-		--privileged $(MULTIARCH_CONTAINER)  \
-		--reset $(VERBOSE_OUTPUT)
-	@docker run --rm --sig-proxy=true            \
+	@docker run --rm --sig-proxy=true        \
 		-v `pwd`/$(OUTPUT_DIR):/build        \
-		-v `pwd`:/src                        \
 		$(COMPILE_IMAGE)                     \
 		/bin/bash -c                         \
-		"tar -xJf /build/dnsmasq.tar.xz      \
+		"curl -sSL $(DNSMASQ_URL) | tar -xJ  \
 			&& cd $(DNSMASQ_VERSION)     \
-			&& CC=$(TRIPLE)-gcc make -j  \
+			&& CC=$(CC) make -j  \
 			&& cp src/dnsmasq /build" $(VERBOSE_OUTPUT)
-endif
-	@cp dnsmasq.conf $(OUTPUT_DIR)
-
-$(DNSMASQ_ARCHIVE):
-	@mkdir -p $(@D)
-	@curl -sSL $(DNSMASQ_URL) > $@
 
 .PHONY: containers
 containers: $(CONTAINER_STAMP)
 
-$(CONTAINER_STAMP): $(BINARY)
-	@echo "container:" $(REGISTRY)/$(IMAGE):$(VERSION)
+$(CONTAINER_STAMP): Dockerfile dnsmasq.conf $(BINARY)
+	@echo "container:" $(REGISTRY)/$(IMAGENAME)-$(ARCH):$(VERSION)
+	@cp dnsmasq.conf $(OUTPUT_DIR)/
+	@sed "s|__BASEIMAGE__|$(BASEIMAGE)|g" Dockerfile > $(OUTPUT_DIR)/Dockerfile
+	@touch $(OUTPUT_DIR)/.emptydir
 	@docker build \
-		-q -t $(REGISTRY)/$(IMAGE):$(VERSION) $(OUTPUT_DIR) > $@
+		-q -t $(REGISTRY)/$(IMAGENAME)-$(ARCH):$(VERSION) $(OUTPUT_DIR) > $@
 
 .PHONY: test
 test: containers
-	@ARCH=$(ARCH) IMAGE=$(REGISTRY)/$(IMAGE):$(VERSION) ./validate.sh
+	@ARCH=$(ARCH) IMAGE=$(REGISTRY)/$(IMAGENAME)-$(ARCH):$(VERSION) ./validate.sh
+
+./manifest-tool:
+	curl -sSL https://github.com/luxas/manifest-tool/releases/download/v0.3.0/manifest-tool > manifest-tool
+	chmod +x manifest-tool
 
+gcr-login:
+ifeq ($(findstring gcr.io,$(PREFIX)),gcr.io)
+	@echo "If you are pushing to a gcr.io registry, you have to be logged in via 'docker login'; 'gcloud docker push' can't push manifest lists yet."
+	@echo "This script is automatically logging you in now."
+	docker login -u oauth2accesstoken -p "$(gcloud auth print-access-token)" https://gcr.io
+endif
 
 .PHONY: push
 push: $(PUSH_STAMP)
 
-$(PUSH_STAMP): $(CONTAINER_STAMP)
-	@echo "pushing  :" $(REGISTRY)/$(IMAGE):$(VERSION)
-	@$(DOCKER) push $(REGISTRY)/$(IMAGE):$(VERSION)
+$(PUSH_STAMP): $(CONTAINER_STAMP) gcr-login
+	@echo "pushing  :" $(REGISTRY)/$(IMAGENAME)-$(ARCH):$(VERSION)
+	@docker push $(REGISTRY)/$(IMAGENAME)-$(ARCH):$(VERSION)
 	@cat $< > $@
 
 .PHONY: clean
```

Comment:
add manifest-tool to clean and any other generated file (I think .emptydir is created as well)